O21.Game/Downloader.fs

```diff
@@ -0,0 +1,175 @@
+module O21.Game.Downloader
+
+open System
+open System.Globalization
+open System.IO
+open System.IO.Compression
+open System.Net.Http
+open System.Net.Http.Handlers
+open System.Security.Cryptography
+open System.Threading.Tasks
+
+open FSharp.Data
+
+open O21.Game.Localization.Translations
+
+type private ContentConfigurationJson = JsonProvider<"content.json">
+
+type ContentConfiguration = {
+    Uri: Uri
+    Sha256: string
+    ContentDirectoryInsideArchive: string
+}
+
+let LoadContentConfiguration(): Task<ContentConfiguration> = task {
+    let! text = File.ReadAllTextAsync Paths.ContentConfigurationFile
+    let config = ContentConfigurationJson.Parse text
+    return {
+        Uri = Uri config.Uri
+        Sha256 = config.Sha256
+        ContentDirectoryInsideArchive = config.ContentDirectoryInsideArchive
+    }
+}
+
+let CheckIfAlreadyLoaded (contentConfig: ContentConfiguration) (dataFolder: string): Task<bool> = task {
+    if Directory.Exists dataFolder then
+        let hashFilePath = Paths.U95ContentHashFile dataFolder
+        if File.Exists hashFilePath then
+            let! actualHash = File.ReadAllTextAsync hashFilePath
+            return actualHash.Trim() = contentConfig.Sha256
+        else return false
+    else return false
+}
+
+let private retryPause = TimeSpan.FromSeconds 5.0
+
+let private BytesToString language =
+    let translation = Translation language // TODO: Get translation from the arguments, not the language
+    let KiB = 1024L
+    let MiB = 1024L * KiB
+
+    function
+    | b when b < KiB -> $"{string b} {translation.BytesAbbreviated}"
+    | b when b < MiB ->
+        let kib = (double b / double KiB).ToString(".##", CultureInfo.InvariantCulture)
+        $"{kib} {translation.KibibytesAbbreviated}"
+    | b ->
+        let mib = (double b / double MiB).ToString(".##", CultureInfo.InvariantCulture)
+        $"{mib} {translation.MebibytesAbbreviated}"
+
+let private DownloadToTempFile contentConfig
+                               (controller: LoadController)
+                               language
+                               progressPart = task {
+    let translation = Translation language // TODO: Get translation from the arguments, not the language
+
+    use handler = new ProgressMessageHandler(new HttpClientHandler(AllowAutoRedirect = true))
```

Comment:
Look for possible resource leaks here: shouldn't we dispose the `HttpClientHandler` as well?

Reply:
No leak.